internal/auth.go

```diff
@@ -0,0 +1,343 @@
+package tfa
+
+import (
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/thomseddon/traefik-forward-auth/internal/provider"
+)
+
+// Request Validation
+
+// Cookie = hash(secret, cookie domain, email, expires)|expires|email
+func ValidateCookie(r *http.Request, c *http.Cookie) (bool, string, error) {
+	parts := strings.Split(c.Value, "|")
+
+	if len(parts) != 3 {
+		return false, "", errors.New("Invalid cookie format")
+	}
+
+	mac, err := base64.URLEncoding.DecodeString(parts[0])
+	if err != nil {
+		return false, "", errors.New("Unable to decode cookie mac")
+	}
+
+	expectedSignature := cookieSignature(r, parts[2], parts[1])
+	expected, err := base64.URLEncoding.DecodeString(expectedSignature)
+	if err != nil {
+		return false, "", errors.New("Unable to generate mac")
+	}
+
+	// Valid token?
+	if !hmac.Equal(mac, expected) {
+		return false, "", errors.New("Invalid cookie mac")
+	}
+
+	expires, err := strconv.ParseInt(parts[1], 10, 64)
+	if err != nil {
+		return false, "", errors.New("Unable to parse cookie expiry")
+	}
+
+	// Has it expired?
+	if time.Unix(expires, 0).Before(time.Now()) {
+		return false, "", errors.New("Cookie has expired")
+	}
+
+	// Looks valid
+	return true, parts[2], nil
+}
+
+// Validate email
+func ValidateEmail(email string) bool {
+	found := false
+	if len(config.Whitelist) > 0 {
+		for _, whitelist := range config.Whitelist {
+			if email == whitelist {
+				found = true
+			}
+		}
+	} else if len(config.Domains) > 0 {
+		parts := strings.Split(email, "@")
+		if len(parts) < 2 {
+			return false
+		}
+		for _, domain := range config.Domains {
+			if domain == parts[1] {
+				found = true
+			}
+		}
+	} else {
+		return true
+	}
+
+	return found
+}
+
+// OAuth Methods
+
+// Get login url
+func GetLoginURL(r *http.Request, nonce string) string {
+	state := fmt.Sprintf("%s:%s", nonce, returnUrl(r))
+
+	// TODO: Support multiple providers
+	return config.Providers.Google.GetLoginURL(redirectUri(r), state)
+}
+
+// Exchange code for token
+
+func ExchangeCode(r *http.Request) (string, error) {
+	code := r.URL.Query().Get("code")
+
+	// TODO: Support multiple providers
+	return config.Providers.Google.ExchangeCode(redirectUri(r), code)
+}
+
+// Get user with token
+
+func GetUser(token string) (provider.User, error) {
+	// TODO: Support multiple providers
+	return config.Providers.Google.GetUser(token)
+}
+
+// Utility methods
+
+// Get the redirect base
+func redirectBase(r *http.Request) string {
+	proto := r.Header.Get("X-Forwarded-Proto")
+	host := r.Header.Get("X-Forwarded-Host")
+
+	return fmt.Sprintf("%s://%s", proto, host)
+}
+
+// // Return url
+func returnUrl(r *http.Request) string {
+	path := r.Header.Get("X-Forwarded-Uri")
```

Comment:
Maybe you should return `X-Forwarded-Prefix` do directly redirect to the initial request.

Reply:
Sorry I'm not quite sure I follow, in https://docs.traefik.io/basics/ it says `X-Forwarded-Prefix` header will be "the stripped prefix path" following a PathPrefixStrip or PathPrefixStripRegex match, so I don't think using this will not be correct in all circumstances.

*However*, using those two matchers can cause issues and there is a fix for this in #39 which I will merge into this branch :)

Reply:
OK, cool. Any eta for the next release ?
Thanks for your good job ! Love it !